libbpf-tools/cpufreq.bpf.c

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "cpufreq.h"
+#include "maps.bpf.h"
+
+static __u32 freqs_mhz[MAX_CPU_NR];
+static struct hist zero;
+struct hist syswide = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct hkey);
+	__type(value, struct hist);
+} hists SEC(".maps");
+
+SEC("tp_btf/cpu_frequency")
+int BPF_PROG(cpu_frequency, unsigned int state, unsigned int cpu_id)
+{
+	if (cpu_id >= MAX_CPU_NR)
+		return 0;
+	freqs_mhz[cpu_id] = state / 1000;
+	return 0;
+}
+
+SEC("perf_event")
+int do_sample(struct bpf_perf_event_data *ctx)
+{
+	u32 freq_mhz, pid = bpf_get_current_pid_tgid();
+	u64 slot, cpu = bpf_get_smp_processor_id();
+	struct hist *hist;
+	struct hkey hkey;
+
+	if (cpu >= MAX_CPU_NR)
+		return 0;
+	freq_mhz = freqs_mhz[cpu];
+	if (!freq_mhz)
+		return 0;
```

Comment:
so until `tp_btf/cpu_frequency` fires, you'll be just ignoring samples? What if the frequency doesn't change at all, you'll be just ignoring data forever? I think you need to pre-initialize freq_mhz array from user-space.

Reply:
Yes, the original idea was to use it only in scenarios like `powersafe`. But in other scenarios there is no output that is really confusing.  I agree with your suggestion, fixed thanks!